src/Mixin.php

```diff
@@ -0,0 +1,539 @@
+     */
+    public static function allBoolean($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert null|scalar $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function nullOrScalar($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert iterable<scalar> $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function allScalar($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert null|object $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function nullOrObject($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert iterable<object> $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function allObject($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert null|resource $value
+     * @param mixed $value
+     * @param string|null $type    type of resource this should be. @see https://www.php.net/manual/en/function.get-resource-type.php
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function nullOrResource($value, $type = null, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert iterable<resource> $value
+     * @param mixed $value
+     * @param string|null $type    type of resource this should be. @see https://www.php.net/manual/en/function.get-resource-type.php
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function allResource($value, $type = null, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert null|callable $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function nullOrIsCallable($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert iterable<callable> $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function allIsCallable($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert null|array $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function nullOrIsArray($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert iterable<array> $value
+     * @param mixed $value
+     * @param string $message
+     * @throws InvalidArgumentException
+     */
+    public static function allIsArray($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert null|iterable $value
+     * @deprecated use "isIterable" or "isInstanceOf" instead
```

Comment:
maybe we should skip deprecated methods?

Reply:
I don't have a strong opinion here:

* from one hand: it makes no sense to generate methods that are based on _currently_ deprecated methods - they haven't been typed, so not a big deal

* from another hand: if some other method is deprecated in the future - its `nullOrX` and `allX` should also be marked as deprecated, not silently removed.

So I personally lean towards generating them.

Reply:
true that: cannot add deprecation markers on `@method`!

This thread can be dismissed